gsy_framework/kafka_communication/kafka_consumer.py

```diff
@@ -31,12 +42,50 @@ def __init__(self, callback):
                       "max_partition_fetch_bytes": KAFKA_MAX_MESSAGE_SIZE_PER_TOPIC,
                       "max_poll_records": KAFKA_MAX_POLL_RECORDS,
                       "consumer_timeout_ms": KAFKA_CONSUMER_TIMEOUT_MS,
-                      "group_id": KAFKA_CONSUMER_GROUP_ID}
+                      "group_id": KAFKA_CONSUMER_GROUP_ID,
+                      "enable_auto_commit": False}
 
         self._consumer = KafkaConsumer(KAFKA_RESULTS_TOPIC, **kwargs)
         self._callback = callback
 
+    @staticmethod
+    def _deserialize_message(msg):
+        return json.loads(decompress(msg.value).decode("utf-8"))
+
+    def _get_topic_partition(self):
+        topic_partitions = self._consumer.assignment()
+        if len(topic_partitions) != 1:
+            # More than one partitions are connected to this consumer, this needs to be
+            # reported as error.
+            logger.error(
+                "Consumer was connected to more than one topic / partition: %s",
+                topic_partitions)
+        return list(topic_partitions)[0]
+
+    def _process_message_with_retries(self, msg):
+        was_processed_correctly = False
+        retry_counter = 0
+        while not was_processed_correctly and retry_counter < RESULTS_SERVICE_RETRY_COUNT:
+            # Try to process the message multiple times. Omit the message if it does not
+            # succeed.
+            try:
+                payload = self._deserialize_message(msg)
+                self._callback(payload)
+                was_processed_correctly = True
+            # pylint: disable=broad-except
+            except Exception as ex:
+                retry_counter += 1
+                logger.error(
+                    "Simulation results service ailed to process incoming message. %s. "
+                    "Traceback: %s", str(ex), traceback.format_exc())
+
     def execute_cycle(self):
+        """Main execution cycle of the Kafka consumer."""
         for msg in self._consumer:
-            payload = json.loads(decompress(msg.value).decode("utf-8"))
-            self._callback(payload)
+            topic_partition = self._get_topic_partition()
+            self._process_message_with_retries(msg)
+            # Commit the message even though it has not been processed, since it failed in
+            # multiple retries.
+            self._consumer.commit({
+                topic_partition: OffsetAndMetadata(msg.offset + 1, None)
```

Comment:
Actually this is more related to the partition and not directly to the message. It actually is a cursor that points to the next message that needs to be sent from this partition. Therefore, in case the message has been processed, we need to increase the offset by 1 message, in order to point to the next message that has not been processed yet.

Reply:
Interesting. IMO this is worth an additional sentence in the comment above the command. What do you think?

Reply:
Let me add some more detail in the comment yes! Fixing now